test/functional/feature_taproot.py

```diff
@@ -0,0 +1,797 @@
+
+def spender_alwaysvalid(spenders, info, comment, err_msg=None, **kwargs):
+    """Mutates the witness when requested, intended for otherwise always true scripts."""
+    spk = info[0]
+    addr = get_taproot_bech32(info)
+
+    def fn(t, i, u, v):
+        return spend_alwaysvalid(t, i, damage=not v, info=info, **kwargs)
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=False, sat_function=fn, err_msg=err_msg, sigops_weight=0))
+
+def spender_two_paths_alwaysvalid(spenders, info, comment, standard, success, failure, err_msg=None):
+    """Allows specifying both a success and failure script without any signatures or additional witness data required."""
+    spk = info[0]
+    addr = get_taproot_bech32(info)
+
+    def fn(t, i, u, v):
+        return spend_no_sig(t, i, u, info, (success if v else failure))
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=standard, sat_function=fn, err_msg=err_msg, sigops_weight=0))
+
+def spender_alwaysvalid_p2sh(spenders, info, comment, standard, script, err_msg=None):
+    """Tests that p2sh-wrapping witness program v1 are always valid, since they are not covered by tapscript."""
+    spk = get_p2sh_spk(info)
+    addr = get_version1_p2sh(info)
+
+    def fn(t, i, u, v):
+        if v:
+            t.vin[i].scriptSig = CScript([info[0]])
+            # Empty control block is only invalid if we apply taproot rules,
+            # which we shouldn't if the spend is wrapped in P2SH
+            t.wit.vtxinwit[i].scriptWitness.stack = [script, bytes()]
+        else:
+            t.vin[i].scriptSig = CScript()
+        return
+
+    spenders.append(Spender(script=spk, address=addr, comment=comment, is_standard=standard, sat_function=fn, err_msg=err_msg, sigops_weight=0))
+
+def nested_script(script, depth):
+    if depth == 0:
+        return script
+    return [nested_script(script, depth - 1), CScript([OP_RETURN])]
+
+UTXOData = namedtuple('UTXOData', 'input,output,spender')
+
+class TaprootTest(BitcoinTestFramework):
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.extra_args = [["-whitelist=127.0.0.1", "-par=1"]]
+
+    def block_submit(self, node, txs, msg, err_msg, cb_pubkey=None, fees=0, sigops_weight=0, witness=False, accept=False):
+
+        # Deplete block of any non-tapscript sigops using a single additional 0-value coinbase output.
+        # It is not impossible to fit enough tapscript sigops to hit the old 80k limit without
+        # busting txin-level limits. We simply have to account for the p2pk outputs in all
+        # transactions.
+        extra_output_script = CScript([OP_CHECKSIG]*((MAX_BLOCK_SIGOPS_WEIGHT - sigops_weight) // WITNESS_SCALE_FACTOR))
```

Comment:
nice, you may get 1 to 3 sigops weight left over by chance, but tests will blow that out regularly.

Reply:
Yeah, I know, but it's a lot harder to test that accurately (it needs creating outputs and spending them using witness spends, ...).

Ideally, we'd also test that if you have one sigop too many, things also fail. Otherwise it could be the case that some of the  sigops estimates in the test are overestimates.